babel/messages/extract.py

```diff
@@ -142,28 +142,74 @@ def extract_from_dir(dirname=None, method_map=DEFAULT_MAPPING,
-                    if callback:
-                        callback(filename, method, options)
-                    for lineno, message, comments, context in \
-                          extract_from_file(method, filepath,
-                                            keywords=keywords,
-                                            comment_tags=comment_tags,
-                                            options=options,
-                                            strip_comment_tags=
-                                                strip_comment_tags):
-                        yield filename, lineno, message, comments, context
-                    break
+            filepath = os.path.join(root, filename).replace(os.sep, '/')
+
+            for (
+                extracted_filename,
+                lineno,
+                message,
+                comments,
+                context,
+            ) in check_and_call_extract_file(
+                    filepath,
+                    method_map,
+                    options_map,
+                    callback,
+                    keywords,
+                    comment_tags,
+                    strip_comment_tags,
+                    dirpath=absname,
+            ):
+                yield extracted_filename, lineno, message, comments, context
+
+
+def check_and_call_extract_file(filepath, method_map, options_map,
+                                callback, keywords, comment_tags,
+                                strip_comment_tags, dirpath=None):
+    """Checks if the given file matches an extraction method mapping, and if so, calls extract_from_file.
+
+    :param filepath: An absolute path to a file that exists.
+    :param method_map: a list of ``(pattern, method)`` tuples that maps of
+                       extraction method names to extended glob patterns
+    :param options_map: a dictionary of additional options (optional)
+    :param callback: a function that is called for every file that message are
+                     extracted from, just before the extraction itself is
+                     performed; the function is passed the filename, the name
+                     of the extraction method and and the options dictionary as
+                     positional arguments, in that order
+    :param keywords: a dictionary mapping keywords (i.e. names of functions
+                     that should be recognized as translation functions) to
+                     tuples that specify which of their arguments contain
+                     localizable strings
+    :param comment_tags: a list of tags of translator comments to search for
+                         and include in the results
+    :param strip_comment_tags: a flag that if set to `True` causes all comment
+                               tags to be removed from the collected comments.
+    :param dirpath: the path to the directory to extract messages from.
+    """
+    if dirpath is None:
+        dirpath = os.getcwd()
```

Comment:
fwiw extract_from_dir(..) above it does the same check. Though I'm fine with having the caller explicitly pass it in.